### PR TITLE
exposed outlierDetection configuration through GTP

### LINKED ISSUE #104: [Enhancement] Make OutlierDetection params configurable
Is there anyway to set outlierDetection? It seems to be hardcoded at the moment, but can we use GTP or other RD to define it?

```yaml
outlierDetection:
      baseEjectionTime: 120s
      consecutive5xxErrors: 10
      interval: 5s
```

### PR DESCRIPTION
**Pull request by shriramsharma**
Fixes #104
Extended the GTP API to expose outlierDetection configuration. 

Example GTP Config:
```
apiVersion: admiral.io/v1alpha1
kind: GlobalTrafficPolicy
metadata:
  name: gtp-failover-config
  namespace: sample
  annotations:
    admiral.io/env: stage
  labels:
    identity: greeting
spec:
  policy:
    - dnsPrefix: default # default is a keyword, alternatively you can use `env` (ex: stage)
      lbType: 1 #0 represents TOPOLOGY, 1 represents FAILOVER
      target:
        - region: us-west-2
          weight: 80
        - region: us-east-2
          weight: 20
      outlier_detection:
        base_ejection_time: 180
        consecutive_gateway_errors: 1000
        interval: 60
```

### CODE REVIEW

**Review by psikka1 — APPROVED** [review]
LGTM 👍

File: admiral/pkg/apis/admiral/model/globalrouting.proto [hunk 1]
```diff
@@ -74,6 +74,19 @@ message TrafficPolicy {
...
+       //REQUIRED: Minimum duration of time in seconds, the endpoint will be ejected
+       int64 base_ejection_time = 1;
+       //REQUIRED: No. of consecutive failures in specified interval after which the endpoint will be ejected
+       uint32 consecutive_gateway_errors = 2;
+       //REQUIRED: Time interval between ejection sweep analysis
+       int64 interval = 3;
+   }
+
+   //OPTIONAL: to configure the outlierDetection in DestinationRule
+   FailoverConfiguration failover_configuration = 5;
```
Comment (nirvanagit) [inline admiral/pkg/apis/admiral/model/globalrouting.proto:88]: Should we call this `outlierDetection` ? Just like in https://istio.io/latest/docs/reference/config/networking/destination-rule/#OutlierDetection

File: admiral/pkg/clusters/handler_test.go [hunk 2]
```diff
@@ -321,58 +320,105 @@ func TestGetOutlierDetection(t *testing.T) {
...
+			locality:         "uswest2",
+			gtpPolicy:        nil,
+			outlierDetection: outlierDetection,
+		},
+		{
+			name:     "Should use the default outlier detection if FailoverConfiguration is nil inside gtpPolicy",
+			se:       se,
+			locality: "uswest2",
+			gtpPolicy: &model.TrafficPolicy{
+				LbType: model.TrafficPolicy_TOPOLOGY,
```
Comment (nirvanagit) [inline admiral/pkg/clusters/handler_test.go:387]: May be use this `topologyGTPPolicy` ?

File: admiral/pkg/apis/admiral/model/globalrouting.proto (hunk 1 above)
Comment (nirvanagit) [reply admiral/pkg/apis/admiral/model/globalrouting.proto:88]: If we do plan to change it to outlierDetection, then will be good to update references of failure configuration or failure_configuration in test cases.

File: admiral/pkg/clusters/handler.go [hunk 3]
```diff
@@ -106,9 +106,21 @@ func getDestinationRule(se *v1alpha32.ServiceEntry, locality string, gtpTrafficP
...
-		BaseEjectionTime:      &types.Duration{Seconds: 300},
+		BaseEjectionTime:         &types.Duration{Seconds: 300},
 		ConsecutiveGatewayErrors: &types.UInt32Value{Value: uint32(50)},
-		Interval:              &types.Duration{Seconds: 60},
+		Interval:                 &types.Duration{Seconds: 60},
+	}
+
+	if gtpTrafficPolicy != nil && gtpTrafficPolicy.FailoverConfiguration != nil {
+		outlierDetection.BaseEjectionTime = &types.Duration{
+			Seconds: gtpTrafficPolicy.FailoverConfiguration.BaseEjectionTime,
```
Comment (nirvanagit) [inline admiral/pkg/clusters/handler.go:116]: What is BaseEjectionTime is not configured?

File: admiral/pkg/clusters/handler.go [hunk 4]
```diff
@@ -106,9 +106,21 @@ func getDestinationRule(se *v1alpha32.ServiceEntry, locality string, gtpTrafficP
...
+		}
+		outlierDetection.ConsecutiveGatewayErrors = &types.UInt32Value{
+			Value: gtpTrafficPolicy.FailoverConfiguration.ConsecutiveGatewayErrors,
```
Comment (nirvanagit) [inline admiral/pkg/clusters/handler.go:119]: What if ConsecutiveGatewayErrors is not configured?

File: admiral/pkg/clusters/handler.go [hunk 5]
```diff
@@ -106,9 +106,21 @@ func getDestinationRule(se *v1alpha32.ServiceEntry, locality string, gtpTrafficP
...
+		}
+		outlierDetection.Interval = &types.Duration{
+			Seconds: gtpTrafficPolicy.FailoverConfiguration.Interval,
```
Comment (nirvanagit) [inline admiral/pkg/clusters/handler.go:122]: Will need to add another check when Interval is not configured.

File: admiral/pkg/clusters/handler.go (hunk 4 above)
Comment (nirvanagit) [reply admiral/pkg/clusters/handler.go:119]: If not then assign default (same for the other two properties)

File: admiral/pkg/clusters/handler.go [hunk 6]
```diff
@@ -106,9 +106,27 @@ func getDestinationRule(se *v1alpha32.ServiceEntry, locality string, gtpTrafficP
 func getOutlierDetection(se *v1alpha32.ServiceEntry, locality string, gtpTrafficPolicy *model.TrafficPolicy) *v1alpha32.OutlierDetection {
 
 	outlierDetection := &v1alpha32.OutlierDetection{
-		BaseEjectionTime:      &types.Duration{Seconds: 300},
+		BaseEjectionTime:         &types.Duration{Seconds: 300},
```
Comment (nirvanagit) [inline admiral/pkg/clusters/handler.go:109]: May be we can define constants for all the defaults and use them in our tests as well.

**Review by nirvanagit — APPROVED** [review]

**Pull request merged**